lib/daru/maths/statistics/vector.rb

```diff
@@ -82,24 +82,38 @@ def factors
         #
         #   dv.max(2) { |a,b| a.size <=> b.size }
         #   #=> ["Jon Starkgaryen","Daenerys"]
+        def max(size=nil, &block)
+          data           = reject_values(*Daru::MISSING_VALUES).to_a
+          range          = size.nil? ? 0 : (0..size-1)
+          default_method = RUBY_VERSION >= '2.2'
+          default_method ? data.max(size, &block) : data.sort(&block).reverse[range]
+        end
+
+        # Returns the maximum value present in the vector, with an optional object block.
+        #
+        # @example
         #
-        #   dv.max(2) { |i| i.size }
+        #   dv = Daru::Vector.new (["Tyrion", "Daenerys", "Jon Starkgaryen"]), index: Daru::Index.new([:t, :d, :j])
+        #   #=>
+        #   #   #<Daru::Vector(3)>
+        #   #       t   Tyrion
+        #   #       d   Daenerys
+        #   #       j   Jon Starkgaryen
+        #
+        #   dv.max_by
+        #   #=> "Tyrion"
+        #
+        #   dv.max_by(2) { |i| i.size }
         #   #=> ["Jon Starkgaryen","Daenerys"]
-        def max(size=nil, &block)
-          data = @data.data.to_a
-          data = if block_given?
-                   if block.parameters.count == 1 # Object block like { |x| x.size }
-                     data.sort_by(&block)
-                   else # Comparative block like { |a,b| a.size <=> b.size }
-                     data.sort(&block)
-                   end
-                 else
-                   data.sort
-                 end
-          size.nil? ? data.last : data[data.count-size..-1].reverse
+        def max_by(size=nil, &block)
+          data           = reject_values(*Daru::MISSING_VALUES).to_a
+          range          = size.nil? ? 0 : (0..size-1)
+          default_method = RUBY_VERSION >= '2.2'
+          default_method ? data.max_by(size, &block || ->(i) { i }) : data.sort_by(&block || ->(i) { i }).reverse[range]
         end
```

Comment:
Second: `->(i) { i }` is `&:itself` → which means entire construct could be shortened to `&(block || :itself)` (requires backports for Ruby < 2.2).

But, third: whether it is a good thing to use "sort by itself" by default? It is a) pretty different from `Array#max_by` behavior and b) useful for nothing (e.g. how `vector.max_by` block-less is better/more readable than `vector.max`?). I believe we should just raise on absent block.